### ISSUE TITLE
Run xattr -r -d com.apple.quarantine App.app on casks on Apple Silicon MacOS

### ISSUE THREAD

**Issue by yesseruser** [Homebrew/brew#17979]
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

On Apple Silicon MacOS (didn't test with x64) unsigned application bundles have a quarantine state and MacOS shows they are corrupted (which they aren't). This can be bypassed by running `xattr -r -d com.apple.quarantine App.app` where App.app is the path to the application bundle.

### What is the motivation for the feature?

Some casks cannot be run without bypassing this.

### How will the feature be relevant to at least 90% of Homebrew users?

Since Apple has fully switched to Apple Silicon on Macs, this is an issue in numerous projects.

### What alternatives to the feature have been considered?

Casks can be requested to do this themselves.

**Comment by MikeMcQuaid** [comment]
> Some casks cannot be run without bypassing this.

Which casks?

Note: we're currently in the process of removing all unsigned casks in Homebrew/homebrew-cask. That said: we should ensure that there's still a way of users to install them from non-official taps.

CC @homebrew/cask for input/thoughts here.

**Comment by carlocab** [comment]
We go out of our way to retain quarantine attributes when we download casks. See https://github.com/Homebrew/brew/blob/90a90b3076910375fa05e99197950cf2846bc853/Library/Homebrew/cask/quarantine.rb.

I think you can just do `brew install --no-quarantine` if you want `xattr -r -d com.apple.quarantine` to be run.

If desired, it may make sense to have a cask DSL that implicitly does `--no-quarantine` for the cask, but I imagine we'd only allow usage of that for external taps.

**Comment by MikeMcQuaid** [comment]
> If desired, it may make sense to have a cask DSL that implicitly does `--no-quarantine` for the cask, but I imagine we'd only allow usage of that for external taps.

Even that: I'm not sure what I'd think about us essentially disabling a macOS security feature here.

**Comment by carlocab** [comment]
> Even that: I'm not sure what I'd think about us essentially disabling a macOS security feature here.

Same, which is why if we do have a cask DSL, I think it should only be for external taps, since I do agree with this:

> That said: we should ensure that there's still a way of users to install them from non-official taps.

`--no-quarantine` is kind of a way (not 100% sure, haven't tested), but requiring that flag to use an external tap isn't very ergonomic. This just tempts users into adding `--no-quarantine` into their `HOMEBREW_CASK_OPTS`, which we don't want.

Here's an alternative suggestion:

Allow `--no-quarantine` to accept a comma-separated list of Cask names. Any cask named there will be installed with the `--no-quarantine` flag set. That way users who really need to install a non-conforming cask won't have to either wrestle with passing `--no-quarantine` when needed (or, more likely, set it in their `HOMEBREW_CASK_OPTS`).

**Comment by MikeMcQuaid** [comment]
> This just tempts users into adding `--no-quarantine` into their `HOMEBREW_CASK_OPTS`, which we don't want.

Agreed. We should probably remove support for this, seems like a bad idea.

I'm not sure I fully understand (and need a @Homebrew/cask person to enlighten me) what the flow looks like if you try to use a quarantined cask without code-signing. Is it just the same as if I'd downloaded it with Safari e.g. requiring a manual "Open With" override?

**Comment by SMillerDev** [comment]
Yes, it's the same as safari. But on Apple silicon the security model doesn't allow for overrides at all AFAIK. So your only choice is removing the quarantine attribute to make macOS think it wasn't downloaded.

**Comment by MikeMcQuaid** [comment]
Yeh, particularly on Apple Silicon: this seems bad and I feel like we shouldn't do that at all.

**Comment by SMillerDev** [comment]
Just to clarify, we shouldn't allow people to avoid the attribute? Or we shouldn't set the attribute because it makes life more difficult for users who want unsigned software in a tap?

**Comment by carlocab** [comment]
> Yeh, particularly on Apple Silicon: this seems bad and I feel like we shouldn't do that at all.

I think that disallowing the use of `--no-quarantine` would mean users would need to run `xattr -r -d` manually themselves if they want to install an unsigned cask (which presumably will just be from an external tap, eventually).

Are we ok with that? I think I would be, but just making sure that that's what we want.

**Comment by MikeMcQuaid** [comment]
> Just to clarify, we shouldn't allow people to avoid the attribute? Or we shouldn't set the attribute because it makes life more difficult for users who want unsigned software in a tap?

Sorry for confusion! We shouldn't allow people to avoid the attribute. We shouldn't be less secure than downloading the same thing from Safari, even for third-party taps, even with an option specified.

> I think I would be, but just making sure that that's what we want.

I think I would be too. 

I think we should aim to preserve the same security model as downloading something from Safari but just improve the automation of this.

I'm also thinking we should look pretty hard at `--no-quarantine` behaviour and decide if we want to keep it around as-is.

**Comment by carlocab** [comment]
> We shouldn't be less secure than downloading the same thing from Safari, even for third-party taps, even with an option specified.

I agree with this. But, once you agree with the above, I think you must conclude the following:

> I'm also thinking we should look pretty hard at `--no-quarantine` behaviour and decide if we want to keep it around as-is.

In particular, unless Safari also has an equivalent of `--no-quarantine`, then that means we should probably remove `--no-quarantine`. CC @Homebrew/maintainers for thoughts.

**Comment by MikeMcQuaid** [comment]
> In particular, unless Safari also has an equivalent of `--no-quarantine`, then that means we should probably remove `--no-quarantine`. CC @Homebrew/maintainers for thoughts.

Yes, I agree this is the natural conclusion.

**Comment by woodruffw** [comment]
With my security hat on: I'm guardedly in favor of removing `--no-quarantine`, at least for first-party casks (and especially if we're removing unsigned apps, since that eliminates the whole problem). I think it would also make sense to remove for third-party casks, **if** we provide a DSL option that third-party casks can use, per @carlocab's point.

(My thinking there is that there's certainly going to be a long tail of 3p casks that can't/won't be signed, and that removing support in `brew` itself (whether via CLI or DSL) means that users will end up fiddling with `xattr` calls, which are easy to get wrong.)

**Comment by p-linnane** [comment]
We should never intentionally disable macOS security features. I have never been a fan of `--no-quarantine`, but I think it makes sense to keep it. Our trust model leans on 'know what you're installing'. brew makes software more accessible to users, and if we need to keep `--no-quarantine` to better support third-parties, then so be it.

macOS Sequoia is already making it more difficult to run unsigned applications. Homebrew/cask has not allowed unsigned apps to be added for well over a year now. I've been keeping my own tap with unsigned applications outside of Homebrew/cask for some time due to this.

To sum up my thoughts:

1. We should not remove quarantine attributes by default
2. We should keep `--no-quarantine` to better support third-party taps
3. We should determine a cut off point where we want all unsigned applications to be removed from Homebrew/cask

**Comment by yesseruser** [comment]
> We should never intentionally disable macOS security features. I have never been a fan of `--no-quarantine`, but I think it makes sense to keep it. Our trust model leans on 'know what you're installing'. brew makes software more accessible to users, and if we need to keep `--no-quarantine` to better support third-parties, then so be it.
> 
> macOS Sequoia is already making it more difficult to run unsigned applications. Homebrew/cask has not allowed unsigned apps to be added for well over a year now. I've been keeping my own tap with unsigned applications outside of Homebrew/cask for some time due to this.
> 
> To sum up my thoughts:
> 
>     1. We should not remove quarantine attributes by default
> 
>     2. We should keep `--no-quarantine` to better support third-party taps
> 
>     3. We should determine a cut off point where we want all unsigned applications to be removed from Homebrew/cask

Does that close the issue?

**Comment by MikeMcQuaid** [comment]
> Our trust model leans on 'know what you're installing'. brew makes software more accessible to users, and if we need to keep `--no-quarantine` to better support third-parties, then so be it.

@p-linnane how do you feel about `HOMEBREW_CASK_OPTS` supporting `--no-quarantine`?

**Comment by p-linnane** [comment]
For the reason you quoted, I would be okay with allowing that.

**Comment by MikeMcQuaid** [comment]
@p-linnane Thanks for clarifying 👍🏻

**Comment by p-linnane** [comment]
@yesseruser I think the takeaway here is that we will allow `--no-quarantine` in `HOMEBREW_CASK_OPTS` so you can set this at a global level and not worry about it anymore. Will open a PR for this soon.

**Comment by p-linnane** [comment]
Turns out we already support this! `export HOMEBREW_CASK_OPTS="--no-quarantine"` should get you what you need @yesseruser.
